### PR TITLE
Only FP datatypes are allowed in buffer_*_format and image load intri…

### PR DESCRIPTION
**Pull request by trenouf**
…nsics

This is required for a forthcoming upgrade to upstream LLVM, which has
started enforcing this restriction.

Change-Id: I0329637eb59d4b4bfeb58553847eec9598b593ff

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 6f4bb15e8ee03eb8f6c94e714cf2bb66d09c7616
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 46251728a7137f399b564b0886bd9bf9888c48ef</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev c740412cd4afe87d79a487bd375996d52360a3a0</li>
<li>LLVM: amd-vulkan-dev 9bc5dd4450a6361faf5c5661056a7ee494fad830</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 365531/365531 (100.0%)</li>
<li>Failed: 0/365531 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 365657/365657 (100.0%)</li>
<li>Failed: 0/365657 (0.0%)</li>
</ul></details>


</ul></details>

File: builder/llpcBuilderImplImage.cpp [hunk 1]
```diff
@@ -864,6 +864,34 @@ static const Intrinsic::ID ImageAtomicIntrinsicTable[][8] =
...
+    Type* pNewTy = pOrigTy;
+    switch (pNewTy->getScalarType()->getIntegerBitWidth())
+    {
+    case 16:
+        pNewTy = Type::getHalfTy(pNewTy->getContext());
+        break;
+    case 32:
+        pNewTy = Type::getFloatTy(pNewTy->getContext());
+        break;
+    case 64:
```
Comment (linqun) [inline builder/llpcBuilderImplImage.cpp]: I don't think our HW support 64bit type? have you hit this case in CTS or some other test?
if not, please add LLPC_NEVER_CALLED()  before your code.

File: builder/llpcBuilderImplImage.cpp [hunk 2]
```diff
@@ -906,14 +930,26 @@ Value* BuilderImplImage::CreateImageLoad(
...
+        if (pTexelTy != pResultTy)
+        {
+            // TFE
+            pIntrinsicDataTy = StructType::get(pTexelTy->getContext(), { pTexelTy, getInt32Ty() });
+        }
+    }
+
     SmallVector<Value*, 16> args;
-    Instruction *pResult = nullptr;
+    Value *pResult = nullptr;
```
Comment (linqun) [inline builder/llpcBuilderImplImage.cpp]: Why do you need change the type from Instruction to Value? it seems it is always instruction in following code.

File: builder/llpcBuilderImplImage.cpp [hunk 3]
```diff
@@ -864,6 +864,34 @@ static const Intrinsic::ID ImageAtomicIntrinsicTable[][8] =
     },
 };
 
+// =====================================================================================================================
+// Convert an integer or vector of integer type to the equivalent (vector of) half/float/double
+static Type* ConvertTypeToFp(
```
Comment (amdrexu) [inline builder/llpcBuilderImplImage.cpp]: ConvertTypeToFp -> ConvertToFloatingPointType

**Comment by trenouf** [comment]
Thanks; I will land this.

File: builder/llpcBuilderImplImage.cpp (hunk 2 above)
Comment (trenouf) [reply builder/llpcBuilderImplImage.cpp]: CreateBitCast returns a Value. The only place it needs to be Instruction is where it is the input to CreateWaterfallLoop.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1c4221a4af8c0877b66c68ac0999abe5d8f5c7dd
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 46251728a7137f399b564b0886bd9bf9888c48ef</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev c740412cd4afe87d79a487bd375996d52360a3a0</li>
<li>LLVM: amd-vulkan-dev 9bc5dd4450a6361faf5c5661056a7ee494fad830</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 365531/365531 (100.0%)</li>
<li>Failed: 0/365531 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 365657/365657 (100.0%)</li>
<li>Failed: 0/365657 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by trenouf** [comment]
Fixed a bug found in extended testing with int image load and waterfall loop.

**Comment by amdvlk-admin** [comment]
### Test summary for commit a277c65642a0bfa35c7fbcbd1a098ce85cff75c3
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 66e78b997748d03d77e1d706c10f1f17e18e5654</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 9b632ef4f132bddc94769702ed8b49efbc39d89c</li>
<li>LLVM: amd-vulkan-dev 9bc5dd4450a6361faf5c5661056a7ee494fad830</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 365531/365531 (100.0%)</li>
<li>Failed: 0/365531 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 365657/365657 (100.0%)</li>
<li>Failed: 0/365657 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**